File: src/commands/use.zig
const std = @import("std");
const yazap = @import("yazap");
const State = @import("../State.zig");
const print_to_user = @import("../print_to_user.zig");

const printError = print_to_user.printError;
const printSuccess = print_to_user.printSuccess;

pub fn useCmd(allocator: std.mem.Allocator, matches: *const yazap.ArgMatches) void {

    // Check if argument correct
    if (matches.getSingleValue("DIR") == null) return;

    const dotfiles_dir = std.fs.cwd().openDir(
        matches.getSingleValue("DIR").?,
        .{},
    ) catch |err| {
        printError("Invalid directory provided: {s}", err);
    };

    // Create state
    const new_state = State{
        .dotfiles_dir = dotfiles_dir,
        .profile_name = matches.getSingleValue("profile"),
        .bootstrap = matches.containsArg("bootstrap"),
    };

    // Check validity
    if (!new_state.checkValidDotfiles()) {
        printError("Could not read dofiles repo: {s}", error.NotTDMRepo);
        return;
    }

    // Save to applications data directory
    const data_dir = getDataDir(allocator) catch |err| {
        printError("Could not access applications data directory: {s}", err);
    };

    new_state.serialize(&data_dir) catch |err| {
        printError("Failed to serialize state: {s}", err);
    };

    // Apply state
    new_state.applyConfig() catch |err| {
        printError("Failed to apply the specified dotfiles: {s}", err);
    };

    printSuccess("Successfully switched dotfiles!\n", .{});
}

fn getDataDir(allocator: std.mem.Allocator) !std.fs.Dir {
    const data_dir: []const u8 = try std.fs.getAppDataDir(allocator, "tdm");
    defer allocator.free(data_dir);

    return std.fs.cwd().makeOpenPath(
        data_dir,
        .{},
    );
}

File: tdm/commands/use.py
import click

from tdm.print_to_user import error, success
from tdm.state import State
from tdm.symlink_manager import SymlinkManager


@click.command()
@click.argument("profile_name")
@click.option("--bootstrap", "-b", is_flag=True, help="Run bootstrap script")
def use(profile_name: str, bootstrap: bool):
    """Switch to a different profile"""
    state = State.current()
    if not state:
        error("No tdm repo deployed.")
        return

    symlink_manager = SymlinkManager.current(state, backup_location=state.backup_location())
    new_state = State(state.repo, profile_name)
    symlink_manager.state = new_state
    symlink_manager.patch()
    new_state.save()
    if bootstrap:
        new_state.run_bootstrap()

    success(f"now using the \033[3m{profile_name}\033[0m profile.")
